algebra/src/fft/domain/domain_selector.rs

```diff
@@ -1,144 +1,46 @@
-use crate::{BasicRadix2Domain, EvaluationDomainImpl, MixedRadix2Domain};
+use crate::{BasicRadix2Domain, EvaluationDomain, MixedRadix2Domain};
 use crate::{
     PrimeField, FpParameters,
 };
 
-/// Wraps EvaluationDomain variants, suited for
-/// different field parameters and domain sizes.
-#[derive(Copy, Clone, Hash, Eq, PartialEq, Debug)]
-pub enum EvaluationDomain<F: PrimeField> {
-    #[doc(hidden)]
-    BasicRadix2Domain(BasicRadix2Domain<F>),
-    #[doc(hidden)]
-    MixedRadix2Domain(MixedRadix2Domain<F>),
-}
-
-impl<F: PrimeField> EvaluationDomainImpl<F> for EvaluationDomain<F> {
-
-    fn new(num_coeffs: usize) -> Option<Self> {
-
-        // Let's assign an index to each domain:
-        // -1: No suitable domain found
-        // 0: BasicRadix2Domain
-        // 1: MixedRadix2Domain
-        // 2: ...
-
-        let mut index = -1;
-        let mut domain_size = std::usize::MAX;
+/// Return the smallest sized and most efficient Evaluation Domain able to support `num_coeffs` size
+pub fn get_best_evaluation_domain<F: PrimeField>(num_coeffs: usize) -> Option<Box<dyn EvaluationDomain<F>>>{
+// Let's assign an index to each domain:
+    // -1: No suitable domain found
+    // 0: BasicRadix2Domain
+    // 1: MixedRadix2Domain
+    // 2: ...
+
+    let mut index = -1;
+    let mut domain_size = std::usize::MAX;
+
+    match BasicRadix2Domain::<F>::compute_size_of_domain(num_coeffs) {
+        Some(size) => {
+            if size < domain_size {
```

Comment:
Is always true that the domain with smaller size is the more efficient one to be chosen?

Reply:
Basic Radix 2 complexity: d*2^d where domain_size = 2^d
Mixed Radix 2 complexity: (2d' + k(q-1)) * (2^d' * q^k), where domain_size = 2^d' * q^k
Where q = 5, d, d' <= 15 and k <= 2 in MNT6-753 case (Please @UlrichHaboeck75 check if the formulas are correct).
One hypothesis would be to write a function called "get_best_evaluation_domain()" in which one computes a rough estimation of the complexity with the above formulas and choose the domain corresponding to the minimum value, and another function "get_smallest_evaluation_domain()" in which one returns the evaluation domain having the smallest size (e.g. to be used in the snark prover for instance)

Reply:
The formula for computational complexity is correct. More generally, if the domain size is $L=p_1^d_1 * ... * p_M^d_M$, then the FFT (in its fastest but memory costly implementation) costs  
$$
Cost = (d_1*(p_1-1)+...+d_M*(p_M-1)) * L 
$$
field multiplications (and the same number of field additions). It is also true that,  given a set of prime factors $p_1,..., p_M$, the computational complexity increases with the size $L=L(d_1,...,d_M)$ of the domain. (This is easily seen from the derivative of $Sum_i d_i (p_i-1)$ in direction orthogonal to log_2(L(d_1,...,d_M)) = const. equals $Sum_i (p_i-1) log_2(p_i)$, which is positive.)